xdsl/xdsl_opt_main.py

```diff
@@ -109,28 +109,23 @@ def run(self):
         """
         Executes the different steps.
         """
-        if not self.args.parsing_diagnostics:
-            modules = self.parse_input()
-        else:
-            try:
-                modules = self.parse_input()
-            except ParseError as e:
-                print(e)
-                exit(0)
-
-        output_list: List[str] = []
-        for module in modules:
-            if not self.args.verify_diagnostics:
-                self.apply_passes(module)
-                output_list.append(self.output_resulting_program(module))
-            else:
+        chunks, file_extension = self.prepare_input()
+        output_stream = self.prepare_output()
+        try:
+            for i, chunk in enumerate(chunks):
                 try:
-                    self.apply_passes(module)
-                except DiagnosticException as e:
-                    print(e)
-
-        contents = "// -----\n".join(output_list)
-        self.print_to_output_stream(contents)
+                    if i > 0:
+                        output_stream.write("// -----\n")
+                    module = self.parse_chunk(chunk, file_extension)
+                    if module is not None:
+                        if self.apply_passes(module):
+                            output_stream.write(self.output_resulting_program(module))
+                    output_stream.flush()
+                finally:
+                    chunk.close()
+        finally:
+            if output_stream is not sys.stdout:
```

Comment:
Maybe we want to be a bit smarter here? We don't want to close any other pipe-like thing (stderr, named pipes, etc, right?)

Reply:
Those mechanism are precisely setting stdout before the process sees it.
Named or unnamed pipes are just stdout, as far as the process knows!